components/edit-collective/sections/SetupTwoFactorAuth.js

```diff
@@ -0,0 +1,335 @@
+import StyledTooltip from '../../StyledTooltip';
+import { H2, H3, P } from '../../Text';
+import { withUser } from '../../UserProvider';
+
+const messages = defineMessages({
+  errorWrongLength: {
+    id: 'TwoFactorAuth.Setup.Form.ErrorLength',
+    defaultMessage: 'Incorrect code length. Please re-enter your 6-digit code.',
+  },
+  errorWrongFormat: {
+    id: 'TwoFactorAuth.Setup.Form.ErrorFormat',
+    defaultMessage: 'Incorrect code format. Please enter only numbers.',
+  },
+  inputLabel: {
+    id: 'TwoFactorAuth.Setup.Form.InputLabel',
+    defaultMessage: 'Please enter your 6-digit code without any dashes.',
+  },
+});
+
+const content = () => (
+  <div>
+    <P fontSize="Caption" lineHeight="Caption">
+      <FormattedMessage
+        id="TwoFactorAuth.Setup.AppInfo"
+        defaultMessage="You can use apps such as Google Authenticator, Authy, 1Password, LastPass, or Microsoft Authenticator to scan the QR code and receive a code to input."
+      />
+    </P>
+  </div>
+);
+
+class SetupTwoFactorAuth extends React.Component {
+  static propTypes = {
+    /** From intl */
+    intl: PropTypes.object.isRequired,
+    /** From graphql query */
+    addTwoFactorAuthToAccount: PropTypes.func.isRequired,
+    data: PropTypes.object,
+    /** From parent component */
+    slug: PropTypes.string,
+    collectiveName: PropTypes.string,
+  };
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      error: null,
+      tokenAdded: null,
+      submitting: false,
+    };
+
+    this.submit = this.submit.bind(this);
+  }
+
+  componentDidMount() {
+    const options = {
+      name: `${this.props.collectiveName}'s Open Collective`,
+      length: 64,
+    };
+    const secret = speakeasy.generateSecret(options);
+    // speakeasy options object does not add issuer as expected so we add it ourselves to the otp url
+    const issuer = '&issuer=Open%20Collective';
+    const fullOTPUrl = secret.otpauth_url + issuer;
+    this.setState({ secret, base32: secret.base32, otpauth_url: fullOTPUrl });
+  }
+
+  async submit(values) {
+    try {
+      // verify QR code
+      const { TOTPCode } = values;
+      const verified = speakeasy.totp.verify({
+        secret: this.state.base32,
+        encoding: 'base32',
+        window: 1,
+        token: TOTPCode,
+      });
+
+      // if not verified, ask the user to try again
+      if (!verified) {
+        this.setState({ error: 'Secret not verified. Please try again.' });
+      }
+
+      // if ok, send secret to backend
+      if (verified) {
+        this.setState({ submitting: true });
+
+        const account = {
+          id: this.props.data.account.id,
+        };
+
+        await this.props
+          .addTwoFactorAuthToAccount({
+            variables: {
+              account,
+              token: this.state.base32,
+            },
+          })
+          .then(() => {
+            confettiFireworks(2000, { zIndex: 3000 });
```

Comment:
I might be grumpy but I feel we have enough confettis. Doesn't hurt for the proof of concept :-)